### PR TITLE
fix: pass package type to loader runner

### LINKED ISSUE #13233: Support for ECMAScript loader (with "type": "module")
## Feature request

**What is the expected behavior?**

If a loader is authored as an ECMAScript module (or if the nearest `package.json` has `"type": "module"`), it would be nice if webpack provided the necessary info to `loader-runner` so that the loader could be loaded.

**What is motivation or use case for adding/changing the behavior?**

It looks like `loader-runner` supports loaders authored with ESM (see webpack/loader-runner#40).  I'm assuming that means that webpack will eventually support ESM loaders.  I'm opening this issue to get confirmation on that or to hear if anybody has looked into it.

**How should this be implemented in your opinion?**

If a loader's `package.json` has `"type": "module"`, webpack should set the `loader.type` to `module` for `loader-runner`.

**Are you willing to work on this yourself?**

Sure (with guidance).

### LINKED ISSUE #15930: [ERR_REQUIRE_ESM]: require() of ES Module when using ES loader
# Bug report

Trying to use a ESM module as a loader, simple repro:

https://github.com/wclr/webpack-esm-loader-issue

**What is the current behavior?**

```
Module build failed (from ./loader.js):
Error [ERR_REQUIRE_ESM]: require() of ES Module ..\webpack-esm-loader-issue\loader.js from D:\Code\webpack-esm-loader-issue\node_modules\loader-runner\lib\loadLoader.js not supported.
Instead change the require of loader.js in ..\webpack-esm-loader-issue\node_modules\loader-runner\lib\loadLoader.js to a dynamic import() which is available in all CommonJS modules.
```

**Other relevant information:**
webpack version:  5.73.0
Node.js version: 16
Operating System: Windows 10

### PR DESCRIPTION
**Pull request by stefanprobst**
this is an attempt to fix #13233, fixes #15930 based on [this comment](https://github.com/webpack/webpack/issues/13233#issuecomment-827459873).

**What kind of change does this PR introduce?**

feature

**Did you add tests for your changes?**

no

**Does this PR introduce a breaking change?**

no

**What needs to be documented once your changes are merged?**

### CODE REVIEW

File: lib/NormalModuleFactory.js [hunk 1]
```diff
@@ -1020,6 +1020,7 @@ If changing the source code is not an option there is also a resolve options cal
 						const parsedResult = identToLoaderRequest(result);
 						const resolved = {
 							loader: parsedResult.loader,
+							type: resolveRequest.descriptionFileData.type,
```
Comment (sokra) [inline lib/NormalModuleFactory.js:1023]: `descriptionFileData` might be unset/missing when no package.json is there

**Review by sokra — CHANGES_REQUESTED** [review]

File: lib/NormalModuleFactory.js [hunk 2]
```diff
@@ -1020,6 +1020,10 @@ If changing the source code is not an option there is also a resolve options cal
 						const parsedResult = identToLoaderRequest(result);
 						const resolved = {
 							loader: parsedResult.loader,
+							type:
```
Comment (sokra) [inline lib/NormalModuleFactory.js:1023]: I think you also want to pass `"module"` when `/\.mjs$/i.test(parsedResult.loader)`.

And when it ends with `.cjs` you don't want to pass `"module"` even if the `descriptionFileData` contains that type

**Review by sokra — CHANGES_REQUESTED** [review]
Please add a test case.

You can put it into `test/cases/loaders`

**Comment by alexander-akait** [comment]
Tests are failed

**Comment by stefanprobst** [comment]
i think i need some handholding how to properly test this. do you have any hints? thanks

**Comment by alexander-akait** [comment]
Wierd, no good message, let's rerun

**Comment by alexander-akait** [comment]
Very strange...

**Comment by otakustay** [comment]
Is there anything I can help to proceed this forward? I tried to run these tests on my local computer and no segment fault is reported under Node 17.4.0 but has another test failed:

```
  ● ConfigTestCases › asset-modules › http-url › http-url should compile

    More errors (10 instead of 6) while compiling than expected:

    The following errors are unexpected:
    moduleIdentifier = /Users/otakustay/Downloads/webpack/test/configCases/asset-modules/http-url/ind...; moduleName = ./index.js; loc = 10:0-106; message = Module not found: Error: connect ECONNREFUSED 0.0.0.0:443
    while fetching https://raw.gi...; moduleId = ./index.js; moduleTrace = []; details = undefined; stack = ModuleNotFoundError: Module not found: Error: connect ECONNREFUSED 0.0.0.0:443
    while fetching https://raw.githubusercontent.com//webpack//webpack//main/CODE_OF_CONDUCT.md
```

File: test/cases/loaders/esm-loader-type/test.filter.js [hunk 3]
```diff
@@ -0,0 +1,5 @@
+module.exports = function(config) {
+	// TODO need fix in v8 https://github.com/nodejs/node/issues/35889
+	// TODO otherwise this test case cause segment fault
+	return false;
```
Comment (vankop) [inline test/cases/loaders/esm-loader-type/test.filter.js:4]: maybe we could filter this spawning another `jest` process and catching exit code.. but it seems overengineering to me..

**Comment by vankop** [comment]
@stefanprobst Thanks for starting PR. I have separated test cases, there is bug in running `import()` in jest currently

**Comment by jablko** [comment]
@vankop Why is this PR preferable to [this simpler one](https://github.com/webpack/webpack/pull/14077)?

**Comment by vankop** [comment]
@jablko here we also respect cjs, mjs extensions and this PR already has test cases.

**Comment by jablko** [comment]
> @jablko here we also respect cjs, mjs extensions

#14077 respects `.cjs` and `.mjs` extensions.

> and this PR already has test cases.

What about revising this PR with the #14077 implementation (thereby retaining this PR's tests)?

**Comment by alexander-akait** [comment]
@stefanprobst sorry for delay, can you rebase?

**Comment by TheLarkInn** [comment]
@stefanprobst we're looking at either next release for first Wednesday of April for this release at latest. I'll review after I cut release today.

**Review by alexander-akait — COMMENTED** [review]
Something wrond with jest again, we really need to figure out how to update it...

File: test/cases/loaders/esm-loader-type/test.filter.js (hunk 3 above)
Comment (vankop) [reply test/cases/loaders/esm-loader-type/test.filter.js:4]: I think yes

**Comment by alexander-akait** [comment]
@stefanprobst Can you comment ESM to pass tests and put `TODO` there, thank you

**Comment by TheLarkInn** [comment]
Congrats @stefanprobst on your first webpack contribution. 🚀 🥇 Thank you!

**Pull request merged**